libvips/foreign/svgload.c

```diff
@@ -364,13 +377,66 @@ typedef VipsForeignLoadSvgClass VipsForeignLoadSvgBufferClass;
 G_DEFINE_TYPE( VipsForeignLoadSvgBuffer, vips_foreign_load_svg_buffer, 
 	vips_foreign_load_svg_get_type() );
 
+#if LIBRSVG_CHECK_FEATURE(SVGZ) && defined(HAVE_ZIP)
+static void *
+vips_zalloc( void *opaque, unsigned items, unsigned size )
+{
+	return( g_malloc0_n( items, size ) );
+}
+
+static void
+vips_zfree( void *opaque, void *ptr )
+{
+	return( g_free( ptr ) );
+}
+#endif
+
 static gboolean
 vips_foreign_load_svg_is_a_buffer( const void *buf, size_t len )
 {
-	char *str = (char *) buf;
+	unsigned char *str = (unsigned char *) buf;
 
 	int i;
 
+#if LIBRSVG_CHECK_FEATURE(SVGZ) && defined(HAVE_ZIP)
+	/* Check for SVGZ gzip signature and inflate.
+	 * Minimum gzip size is 18 bytes, starting with 1F 8B.
+	 */
+	if( len >= 18 && str[0] == 0x1f && str[1] == 0x8b ) {
+		z_stream zs;
+		size_t opos = 0;
+		unsigned char obuf[224];
+
+		zs.zalloc = (alloc_func) vips_zalloc;
+		zs.zfree = (free_func) vips_zfree;
+		zs.opaque = Z_NULL;
+		zs.next_in = str;
+		zs.avail_in = len;
+
+		if( inflateInit2(&zs, 15 | 32) != Z_OK ) {
+			vips_error( "svgload",
+				"%s", _( "Zlib init failed" ) );
+			return( -1 );
+		}
+
+		do {
+			zs.avail_out = sizeof(obuf) - opos;
+			zs.next_out = obuf + opos;
+			if( inflate(&zs, Z_NO_FLUSH) < Z_OK ) {
+				vips_error( "svgload",
+					"%s", _( "Zlib inflate failed" ) );
+				return( -1 );
+			}
+			opos = sizeof(obuf) - zs.avail_out;
+		} while( opos < sizeof(obuf) && zs.avail_out == 0 );
```

Comment:
Is this test right? avail_out will be zero exactly when opos == sizeof(obuf), so they are testing the same thing. 

Should it be `while( output_not_full && input_available )`? which would be `while( opos < sizeof(obuf) && zs.avail_in > 0 )`, I think.